src/AutomaticScheduling.cpp

```diff
@@ -0,0 +1,268 @@
+    map<string, vector<Function> > call_to_caller;
+
+    void construct(Function f, set<string> &visited) {
+        visited.insert(f.name());
+        map<string, Function> calls = find_direct_calls(f);
+        for (map<string, Function>::iterator I = calls.begin(), E = calls.end(); I != E; ++I) {
+            string name = I->first;
+            Function func = I->second;
+            name_to_func[name] = func;
+            call_to_callee[f.name()].push_back(func);
+            call_to_caller[func.name()].push_back(f);
+
+            if (visited.find(I->first) == visited.end()) {
+                construct(I->second, visited);
+            }
+        }
+    }
+
+    void transitive_calls_helper(Function f, vector<Function> &result, set<string> &visited) {
+        visited.insert(f.name());
+        const vector<Function> &dir_calls = calls(f);
+        for (vector<Function>::const_iterator I = dir_calls.begin(), E = dir_calls.end(); I != E; ++I) {
+            Function call = *I;
+            if (visited.find(call.name()) == visited.end()) {
+                result.push_back(call);
+                transitive_calls_helper(call, result, visited);
+            }
+        }
+    }
+};
+
+/** Recursively reset the schedules for the given function and all of
+ * the functions it calls. */
+class ResetSchedules {
+public:
+    ResetSchedules(Function root) {
+        reset_schedule(root);
+        CallGraph cg(root);
+        vector<Function> calls = cg.transitive_calls(root);
+        for (std::vector<Function>::iterator I = calls.begin(), E = calls.end(); I != E; ++I) {
+            reset_schedule(*I);
+        }
+    }
+    
+private:
+    /** Hackish way of resetting a function schedule to the
+     * default. Mostly yanked from Function::define(). */
+    void reset_schedule(Function f) {
+        Schedule &olds = f.schedule();
+        std::vector<Bound> oldbounds(olds.bounds().begin(), olds.bounds().end());
+        f.schedule() = Schedule();
+        Schedule &s = f.schedule();
+        s.bounds().insert(s.bounds().begin(), oldbounds.begin(), oldbounds.end());
+
+        for (size_t i = 0; i < f.args().size(); i++) {
+            Dim d = {f.args()[i], ForType::Serial, DeviceAPI::Parent, true};
+            f.schedule().dims().push_back(d);
+            f.schedule().storage_dims().push_back(f.args()[i]);
+        }
+
+        // Add the dummy outermost dim
+        {
+            Dim d = {Var::outermost().name(), ForType::Serial, DeviceAPI::Parent, true};
+            f.schedule().dims().push_back(d);
+        }
+    }
+};
+
+/** Return min/max bounds for each dimension of the given function
+ * across all callsites. If the bounds cannot be computed, returns an
+ * empty vector. */
+vector<Interval> get_function_bounds(Function f, CallGraph &cg) {
+    vector<Interval> result;
+    result.resize(f.dimensions());
+    vector<bool> initialized;
+    initialized.resize(f.dimensions(), false);
+    vector<Function> callers = cg.callers(f);
```

Comment:
Could this be a reference/const reference to the vector in cg? I doubt its an issue now, but we've had problems in the past where passing things by value eventually added up to significant performance issues. Ditto this in a few more places below.